File: ArkheOS/src/arkhe/polyglot/ucd.R
# ucd.R – Universal Coherence Detection in R

ucd_analyze <- function(data) {
    if (nrow(data) > 1) {
        corr_mat <- abs(cor(t(data)))
        C <- mean(corr_mat[lower.tri(corr_mat)])
    } else {
        C <- 0.5
    }
    F <- 1.0 - C
    return(list(C = C, F = F, conservation = (abs(C + F - 1.0) < 1e-10)))
}

data <- matrix(c(1,2,3,4, 2,3,4,5, 5,6,7,8), nrow=3, byrow=TRUE)
print(ucd_analyze(data))

File: ArkheOS/src/arkhe/polyglot/ucd.m
% ucd.m – Universal Coherence Detection in MATLAB/Octave

function result = ucd_analyze(data)
    [nvars, nobs] = size(data);
    if nvars > 1
        corr_mat = abs(corrcoef(data'));
        tri = triu(corr_mat, 1);
        C = mean(tri(tri~=0));
    else
        C = 0.5;
    end
    F = 1.0 - C;
    result = struct('C', C, 'F', F, 'conservation', (abs(C + F - 1.0) < 1e-10));
end

data = [1 2 3 4; 2 3 4 5; 5 6 7 8];
disp(ucd_analyze(data))
